### PR TITLE
Release cs_main before calling ProcessNewBlock, or processing headers (cmpctblock handling)

### PR DESCRIPTION
**Pull request by sdaftuar**
If a compact block reconstructed successfully without any getblocktxn roundtrip, then we were calling ProcessNewBlock with cs_main held.

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
Arg.
utACK 19d7db214da39765c048d1350ef368aedf858ca8

File: src/main.cpp [hunk 1]
```diff
@@ -5880,6 +5886,10 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
                 return ProcessMessage(pfrom, NetMsgType::HEADERS, vHeadersMsg, nTimeReceived, chainparams, connman);
             }
         }
+        } // cs_main
+
+        if (fBlockReconstructed)
+            return ProcessMessage(pfrom, NetMsgType::BLOCKTXN, blockTxnMsg, nTimeReceived, chainparams, connman);
```
Comment (ryanofsky) [inline src/main.cpp]: Looks like cs_main is still held during the ProcessMessage call above on line 5886. Maybe deserves a comment if this is intentional.

Comment (TheBlueMatt) [inline src/main.cpp]: Oh, actually this is not intentional, this is a separate bug.

Comment (sdaftuar) [inline src/main.cpp]: Added a commit that addresses this one as well.

**Comment by morcos** [comment]
@fanquake when you get a chance, this no longer needs backporting (see above 9259 which was closed)

**Comment by TheBlueMatt** [comment]
re-utACK 8c2e1c072a8c87b6fc1832e62f55e275284e7a3a (without backport)

**Comment by MarcoFalke** [comment]
utACK, needs rebase.

**Comment by morcos** [comment]
utACK 0f1190f

**Comment by gmaxwell** [comment]
still needs rebase.

**Comment by sdaftuar** [comment]
Rebased.

**Comment by TheBlueMatt** [comment]
Needs rebase again :/

**Comment by sdaftuar** [comment]
Rebased.

**Comment by sipa** [comment]
utACK bd02bddb938fa7ddf4e5dc7c3ce2a787feba848a

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin/bitcoin/commit/bd02bddb938fa7ddf4e5dc7c3ce2a787feba848a

**Comment by morcos** [comment]
lightly tested reACK bd02bdd

**Pull request merged**